seed_data/lib.py

```diff
@@ -170,102 +165,94 @@ def set_edx_key(cls, obj, course_run):
         """Sets the edX course id, wherever it exists in the cached object data property"""
         raise NotImplementedError
 
-    def set_data(self, obj, course_run, **kwargs):
-        """Properly sets values on the cached edX object data property"""
+    def build_data_property(self, course_run, **kwargs):
+        """Builds the value of the data property for a cached edX object"""
         raise NotImplementedError
 
     def set_or_create(self, course_run, **kwargs):
         """Ensures that a cached edX object exists, then properly sets values on the data property"""
+        data = self.build_data_property(course_run, **kwargs)
+        obj, created = self.get_or_create(course_run, data=data)
+        if not created:
+            obj.data = data
+            obj.save()
+        CachedEdxUserData.update_cache_last_access(self.user, self.cache_type, timestamp=datetime.utcnow())
+        return obj
+
+    def set_blank(self, course_run):
+        """Ensures that a cached edX object exists and that it has data=None"""
         obj = self.get_or_create(course_run)
-        obj.data = obj.data or {}
-        self.set_data(obj, course_run, **kwargs)
+        if obj.data:
+            obj.data = {}
+            obj.save()
         return obj
 
     def get(self, course_run):
         """Gets a cached edX object for a given User and CourseRun"""
         return self.model_cls.objects.get(user=self.user, course_run=course_run)
 
-    def get_or_create(self, course_run):
+    def get_or_create(self, course_run, data=None):
         """Gets or creates a cached edX object for a given User and CourseRun"""
-        try:
-            return self.model_cls.objects.get_or_create(user=self.user, course_run=course_run)[0]
-        except IntegrityError:
-            return self.model_cls.objects.create(user=self.user, course_run=course_run)
+        defaults = dict(user=self.user, course_run=course_run, data=data)
```

Comment:
`get_or_create` should use the `user` / `course_run` values passed to it so `defaults` can just include `data`.